Refactor: extract reusable core for consumer libraries

TxHawks opened the pull request:
<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:
~~This PR adds a function that logically converts properties and values based on the flow-direction context, in line with the CSSWG's [Logical Properties and Values Level 1 proposal](https://www.w3.org/TR/css-logical-1/).~~
This PR extracts the core functionality of rtl-css-js and makes it available separately, so that it can be consumed by other libraries offering other approaches to bidirectional style authoring.

<!-- Why are these changes necessary? -->
**Why**:
rtl-css-js does a great job in converting styles that were written in an `ltr` context so that they work in an `rtl` one. However, this mode of authoring forces us to be explicit about directions, and implicitly convert them later, somewhat magically (usually in a plugin). Such automatic conversion of `left` to `right`, `ltr` to `rtl`, etc. can quickly become confusing and hard to author, especially in interfaces which are not primarily in one direction, and in which the opposite direction is not somewhat of an afterthought.

CSS is now slowly starting to progress towards allowing the authoring of flow-realtive styles, with a CSSWG proposal in place. The most basic of features, such as `text-align: start` are already supported in some browsers. Others, such as `margin-start`, have some prefixed support, but most of the proposed features do not currently have any browser support.

This pull-request was motivated by this short discussion in [rofrischmann/fela#344 (comment)](https://github.com/rofrischmann/fela/issues/344#issuecomment-323451514), and aims to provide a way to author styles in line with the CSSWG proposal (and expand on it a little bit), while still ensuring backwards compatibility with the way thing were done before.

Since @kentcdodds is uncomfortable with including properties and values that have not yet been finalized in fear of breaking changes down the road, extracting and exposing the core functionality enables the writing external libraries based on this one, that can offer other approaches.

<!--
This pull-request will enable write code like this:
```js
{
  margin: 'logical 0 20px 2px 10px',
  insetInlineStart: '10px', // Ugly, I know, but that's the spec
  end: '5px', // Just because the above is so ugly
  paddingStart: '1rem'
}
```

which will be converted into:
```js
// rtl context
{
  margin: '0 10px 2px 20px',
  right: '10px',
  left: '5px',
  paddingRight: '1rem'
}

// ltr context
{
  margin: '0 20px 2px 10px',
  left: '10px',
  right: '5px',
  paddingleft: '1rem'
}
```
-->

<!-- How were these changes implemented? -->
**How**:
Break `index.js` into smaller files so that most logic can be reused. ~~between `toRtl` and `flowRelative`, add tests to ensure everything is 100% backwards compatible and that all the new functionality works as expected.~~

<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [ ] _Documentation_ <-- Wanted to gauge interest first, will take care of it if relevant.
- [x] Tests
- [ ] _Ready to be merged_ <-- Missing docs and updated build.
- [ ] Added myself to contributors table 

**Note**:
~~This is a big change (well, addition, really). rtl-css-js is a really great library, already doing what it does really well. I do think that this is a nice, optional addition to the set of tools offered by the library, but I completely understand if this suggestion is not something you would like implemented here. If this is indeed the case, I will happily maintain it separately, obviously noting its origins in this project. Otherwise, I am willing to commit to helping maintain this part of the functionality.~~

kentcdodds commented:
Wow! This is quite the pull request! Thanks for all the work here! I'm afraid that I don't follow the CSS standardization process very closely. Can you tell me what it means for Logical Properties and Values to be a Level 1 proposal? Do you know whether that's similar to a stage-1 proposal from the TC39?

I'm hesitant to add non-standard CSS rules to this library. If the standard changes, then we'd have to update things which would be a breaking change. I'd like to avoid that churn in the core library. However, I think adding a plugin system to support that kind of thing would be valuable. Would you be interested in changing your PR to create a plugin system and then creating another package to act as a plugin for `rtl-css-js`? Then folks can use your plugin with `rtl-css-js` and when it's standardized we can pull the plugin into the core. What do you think?

TxHawks commented:
Level one is actually unrelated to the proposal's progression. After CSS3, the spec was broken into modules, each having it's own level, meaning something akin to a revision, like CSS3 was supposedly the third revision of the standard.
The stages are actually named. Here is an [article](https://css-tricks.com/css-standards-process/) explaining them. The CSS Logical Properties and Values spec is currently at the second stage - a published working draft. It is definitely unstable and there is no guarantee that it will ever actually be implemented (although, as mentioned in the pull request, parts of it are already implemented in so.e browsers. See [relevant caniuse entry](http://caniuse.com/#feat=css-logical-props), though it incorrectly marks it as unofficial.
If naming does change at some point, we don't have to break compatibility. We could instead leave support for the existing, outdated names and add support for the revised ones.

If you feel better having this as a plugin, how should you go about doing that? I mean what sort of plugin system do you have in mind? Just expose the reusable parts as importable?

kentcdodds commented:
Thanks for that explanation. I do still think that this would be best as a plugin for now with a new plugin system.

> If you feel better having this as a plugin, how should you go about doing that? I mean what sort of plugin system do you have in mind? Just expose the reusable parts as importable?

Is it ok if I leave that up to your creative mind? My wife and I just had a baby (8 hours ago!) so I don't have any cycles in my time or brain for it right now. I hope you understand. Thanks!

TxHawks commented:
First of all, congratulations. You must be the most devoted open source maintainer in history, looking at pull-requests 8 hours after having a baby.

I can't really imagine it as a plugin per-se, since what it really does, is replace the core functionality, not plug into or enhance it. 
I think that setting a provider-consumer system is a better fit here, and very easy to do simply by exposing the reusable parts as `core`, which could be imported and used by other libs:

```js
import { propertyValueConverters, } from 'rtl-css-js/core';

// Consumer code here
```
I'll go ahead and change the PR, and as soon as it is merged and published, will publish a separate  package with the new behavior depending on `rtl-css-js`.

Maybe we can add an Ecosystem section to the readme, to both help with discoverability and encourage people to create consumers

kentcdodds commented:
Sounds perfect! I'll get to your other pull request as soon as I am able. Luckily this repo has automated releases when code is merged into master so you should get your updates immediately 👍

TxHawks commented:
I've changed up the PR and code to only extract the core functionality out so that it can be consumed by other libraries and squashed all commits together. Documentation to follow very soon.

Only problem right now is that I'm not sure how to configure the build to also include a `core` file that can be consumed by libraries (from `src/core/index.js`). Can you point me in the right direction?

TxHawks commented:
Another option is to publish `core` as a seperate package from a seperate repository and have rtl-css-js depend on it like any other consumer.

If you like this idea better, just open the provider repository (maybe `bidi-css-js-provider`?) and I'll take care of setting it up

kentcdodds commented:
Let's not do a new package. I'll just help get this working and we'll let people import `core`. I don't want the overhead of another package.

kentcdodds commented:
Alrighty! So I've got it working well. See the docs for how to import/require things!

I'll merge this as soon as travis finishes successfully :)

kentcdodds commented:
Actually, I'd prefer to have another person review and merge this... @xymostech or @yzimet? Could you give this a review?

TxHawks commented:
Thanks! 
Didn't want to create work for you, but didn't manage to figure out how to reconfigure the build myself

yzimet commented on src/index.js:
```diff
@@ -20,143 +29,16 @@ const valuesToConvert = arrayToObject([
+// A situation we're accepting here:
+// url('/left/right/rtl/ltr.png') will be changed to url('/right/left/ltr/rtl.png')
+// Definite trade-offs here, but I think it's a good call.
+const bgImgDirectionRegex = new RegExp(
```
Could these RegExps (as well as those defined in `propertyValueConverters.transform`) be defined in a separate constants file? Then the individual converter methods can import them as needed, instead of passing them through on line 113 below.

TxHawks replied:
They need to be passed through, because different consumers will use different RegExps.
They can still be defined in a separate constants file.

The RegExp in`propertyValueConverters.transform` was left in place because I couldn't see how it may become implementation specific.

yzimet commented on src/index.js:
```diff
@@ -224,7 +106,13 @@ function getValueDoppelganger(key, originalValue) {
+    newValue = valueConverter({
+      value: importantlessValue,
+      valuesToConvert,
+      isRtl: true,
```
could you remind me, when is `isRtl` false?

TxHawks replied:
In the current implementation - never. However, the idea behind this refactoring is to allow other libs to use the core of rtl-css-js for other approaches to authoring bidirectional styels, specifically the one outlined in the [CSSWG's Logical Properties and Values Level 1 proposal](https://www.w3.org/TR/css-logical-1/).
In that scenario, an author will write `paddingStart`, which, depending on whether `isRtl` is `true` or `false`, will be converted to `paddingRight` or `paddeingLeft`, respectively.

TxHawks commented:
Yep, all changes are non-breaking. All tests pass without modification

kentcdodds commented:
Should be released automatically soon. Thanks so much!

kentcdodds commented:
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the CONTRIBUTING.md file (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)
